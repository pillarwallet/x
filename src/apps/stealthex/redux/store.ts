import { configureStore, TypedStartListening } from '@reduxjs/toolkit';
import exchangeReducer from './reducers/exchange';
import { exchange } from '../lib/backend/api';

export const store = configureStore({
    reducer: {
        exchange: exchangeReducer,
        [exchange.reducerPath]: exchange.reducer
    },
    middleware: getDefaultMiddleware =>
        getDefaultMiddleware({ serializableCheck: false }).concat(exchange.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppStartListening = TypedStartListening<RootState, AppDispatch>; 