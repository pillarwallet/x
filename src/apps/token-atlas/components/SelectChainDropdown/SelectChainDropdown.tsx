// api
import { useRecordPresenceMutation } from '../../../../services/pillarXApiPresence';

// hooks
import useTransactionKit from '../../../../hooks/useTransactionKit';
import { useAppDispatch, useAppSelector } from '../../hooks/useReducerHooks';

// reducer
import {
  setIsAllChainsVisible,
  setIsSelectChainDropdownOpen,
  setSelectedChain,
} from '../../reducer/tokenAtlasSlice';

// types
import { ChainType } from '../../types/types';

// utils
import { getChainName } from '../../../../utils/blockchain';

// components
import Body from '../Typography/Body';

// images
import ArrowDown from '../../images/arrow-down.svg';

type SelectChainDropdownProps = {
  options: number[];
  className?: string;
};

const SelectChainDropdown = ({
  className,
  options,
}: SelectChainDropdownProps) => {
  /**
   * Import the recordPresence mutation from the
   * pillarXApiPresence service. We use this to
   * collect data on what asset's chain is being selected
   */
  const [recordPresence] = useRecordPresenceMutation();

  const { walletAddress: accountAddress } = useTransactionKit();

  const dispatch = useAppDispatch();
  const isSelectChainDropdownOpen = useAppSelector(
    (state) => state.tokenAtlas.isSelectChainDropdownOpen as boolean
  );
  const selectedChain = useAppSelector(
    (state) => state.tokenAtlas.selectedChain as ChainType
  );

  const allChainsOption: ChainType = {
    chainId: 0,
    chainName: 'all',
  };

  const allOptions = [allChainsOption.chainId, ...options];

  const handleDropdownToggle = () => {
    dispatch(setIsSelectChainDropdownOpen(!isSelectChainDropdownOpen));
  };

  // this will filter the tokens by chain id
  const handleSelectChainId = (option: number) => {
    dispatch(
      setSelectedChain({
        chainId: Number(option),
        chainName: option === 0 ? 'all' : getChainName(option),
      })
    );
    if (accountAddress) {
      recordPresence({
        address: accountAddress,
        action: 'app:tokenAtlas:chainSelect',
        value: { chainId: option },
      });
    }
    dispatch(setIsSelectChainDropdownOpen(false));
    dispatch(setIsAllChainsVisible(false));
  };

  return (
    <div
      id="token-atlas-select-chain-dropdown"
      className={`relative ${className}`}
      data-testid="select-chain-dropdown"
    >
      <button
        type="button"
        className={`flex justify-between items-center w-full h-full p-4 bg-medium_grey rounded focus:outline-none ${
          isSelectChainDropdownOpen && 'border-b border-dark_grey'
        }`}
        onClick={handleDropdownToggle}
      >
        <Body className="text-[17px] mobile:text-[15px] font-medium leading-5 line-clamp-1 capitalize">
          {selectedChain.chainId !== 0
            ? selectedChain.chainName
            : 'Select chain'}
        </Body>
        <img
          src={ArrowDown}
          alt="arrow-down"
          className={`transform ${
            isSelectChainDropdownOpen ? 'rotate-180' : ''
          }`}
        />
      </button>
      {isSelectChainDropdownOpen && (
        <div
          id="token-atlas-select-chain-list"
          className="absolute top-full left-0 bg-medium_grey rounded-b w-full max-h-[400px] border-[1px] border-dark_grey overflow-y-auto capitalize z-50"
        >
          <ul>
            {allOptions.map((option) => (
              // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions
              <li
                key={option}
                id={`token-atlas-select-chain-dropdown-${option}`}
                className="px-4 py-2 cursor-pointer hover:bg-light_grey hover:text-dark_grey border-b border-dark_grey last:border-none"
                onClick={() => handleSelectChainId(option)}
              >
                <Body className="text-base mobile:text-sm font-medium">
                  {option === 0 ? 'all' : getChainName(option)}
                </Body>
              </li>
            ))}
          </ul>
        </div>
      )}
    </div>
  );
};

export default SelectChainDropdown;
