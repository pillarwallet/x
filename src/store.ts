/* eslint-disable @typescript-eslint/no-use-before-define */
// Core
import {
  Middleware,
  Reducer,
  combineReducers,
  configureStore,
  createDynamicMiddleware,
} from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';

// Services
import swapSlice from './apps/the-exchange/reducer/theExchangeSlice';
import tokenAtlasSlice from './apps/token-atlas/reducer/tokenAtlasSlice';
import { pillarXApiPresence } from './services/pillarXApiPresence';
import { pillarXApiTransactionsHistory } from './services/pillarXApiTransactionsHistory';
import { pillarXApiWaitlist } from './services/pillarXApiWaitlist';

// Initialisation
const dynamicMiddleware = createDynamicMiddleware();
const middlewareReducers: { [key: string]: Reducer } = {};

/**
 * @name addReducer
 * @description addReducer is a function that allows anyone
 * working with PillarX to add reducers to the store
 * from their own application.
 *
 * @param newReducer
 */
export const addReducer = (newReducer: {
  reducerPath: string;
  reducer: Reducer;
}) => {
  middlewareReducers[newReducer.reducerPath as string] = newReducer.reducer;
  store.replaceReducer(combineReducers(middlewareReducers));
};

/**
 * @name addMiddleware
 * @description addMiddleware is a function that allows anyone
 * working with PillarX to add RTK middleware to the store
 * from their own application.
 *
 * @param newMiddleware
 */
export const addMiddleware = (newMiddleware: {
  reducerPath: string;
  reducer: Reducer;
  middleware: Middleware;
}) => {
  middlewareReducers[newMiddleware.reducerPath as string] =
    newMiddleware.reducer;
  dynamicMiddleware.addMiddleware(newMiddleware.middleware);

  store.replaceReducer(combineReducers(middlewareReducers));
};

/**
 * Export a store from RTK
 */
export const store = configureStore({
  // Empty reducer for now - the addMiddleware function
  // below will dynamically regenerate the reducers required
  // from the middleware functions.
  reducer: {},
  // Adding the api middleware enables caching, invalidation, polling,
  // and other useful features of `rtk-query`.
  // Note: here we have added dynamicMiddleware.middleware
  // which instructs the store to use the middleware that we
  // will add later.
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().prepend(dynamicMiddleware.middleware),
});

/**
 * Whilst we're here we'll immediately mount the
 * PillarXApi middleware. This will contain all the
 * core API functionality for the PillarX application.
 */
addMiddleware(pillarXApiWaitlist);
addMiddleware(pillarXApiPresence);
addMiddleware(pillarXApiTransactionsHistory);
addReducer(swapSlice);
addReducer(tokenAtlasSlice);

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

// // Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
