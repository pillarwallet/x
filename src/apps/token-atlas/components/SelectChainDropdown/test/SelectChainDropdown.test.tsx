import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

// redux store
import { Provider } from 'react-redux';
import { store } from '../../../../../store';

// reducer
import {
  setIsSelectChainDropdownOpen,
  setSelectedChain,
} from '../../../reducer/tokenAtlasSlice';

// components
import SelectChainDropdown from '../SelectChainDropdown';

const mockOptions = [1, 137, 100];

describe('<SelectChainDropdown />', () => {
  beforeEach(() => {
    store.dispatch(setSelectedChain({ chainId: 0, chainName: 'all' }));
    store.dispatch(setIsSelectChainDropdownOpen(false));
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <SelectChainDropdown options={mockOptions} />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('toggles dropdown visibility on button click', () => {
    render(
      <Provider store={store}>
        <SelectChainDropdown options={mockOptions} />
      </Provider>
    );

    // Initially dropdown should be closed
    expect(screen.queryByRole('list')).not.toBeInTheDocument();

    // Click to open dropdown
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByRole('list')).toBeInTheDocument();

    // Click to close dropdown
    fireEvent.click(screen.getByRole('button'));
    expect(screen.queryByRole('list')).not.toBeInTheDocument();
  });

  it('displays the correct chain name when an option is selected', () => {
    render(
      <Provider store={store}>
        <SelectChainDropdown options={mockOptions} />
      </Provider>
    );

    // Open the dropdown
    fireEvent.click(screen.getByRole('button'));

    // Select the first chain option
    fireEvent.click(screen.getByText('Ethereum'));

    // Check if the selected chain is displayed correctly
    expect(screen.getByText('Ethereum')).toBeInTheDocument();
    expect(store.getState().tokenAtlas.selectedChain.chainId).toBe(1);
  });

  it('closes the dropdown when an option is selected', () => {
    render(
      <Provider store={store}>
        <SelectChainDropdown options={mockOptions} />
      </Provider>
    );

    // Open the dropdown
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByRole('list')).toBeInTheDocument();

    // Select an option
    fireEvent.click(screen.getByText('Ethereum'));

    // Dropdown should be closed
    expect(screen.queryByRole('list')).not.toBeInTheDocument();
  });

  it('displays the correct number of options including "all"', () => {
    render(
      <Provider store={store}>
        <SelectChainDropdown options={mockOptions} />
      </Provider>
    );

    // Open the dropdown
    fireEvent.click(screen.getByRole('button'));

    // Check if the dropdown contains all options including "all"
    const options = screen.getAllByRole('listitem');
    expect(options).toHaveLength(mockOptions.length + 1);
    expect(screen.getByText('all')).toBeInTheDocument();
    expect(screen.getByText('Ethereum')).toBeInTheDocument();
  });
});
