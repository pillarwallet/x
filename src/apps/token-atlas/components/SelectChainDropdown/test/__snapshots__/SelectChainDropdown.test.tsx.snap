// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SelectChainDropdown /> > renders correctly and matches snapshot 1`] = `
<div
  className="relative undefined"
  data-testid="select-chain-dropdown"
  id="token-atlas-select-chain-dropdown"
>
  <button
    className="flex justify-between items-center w-full h-full p-4 bg-medium_grey rounded focus:outline-none false"
    onClick={[Function]}
    type="button"
  >
    <p
      className="text-[13px] font-semibold text-[17px] mobile:text-[15px] font-medium leading-5 line-clamp-1 capitalize"
    >
      Select chain
    </p>
    <img
      alt="arrow-down"
      className="transform "
      src="data:image/svg+xml,%3csvg%20width='14'%20height='14'%20viewBox='0%200%2014%2014'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M11.6199%205.22083L7.81655%209.02417C7.36738%209.47333%206.63238%209.47333%206.18322%209.02417L2.37988%205.22083'%20stroke='%23A3A3A3'%20stroke-width='2'%20stroke-miterlimit='10'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3c/svg%3e"
    />
  </button>
</div>
`;
